### PR TITLE
BidHandshakeCycle Management

### PR DESCRIPTION
**Pull request by mjoyce91**
Merge #1568 first

Use with https://github.com/MetaPhase-Consulting/State-TalentMAP-API/pull/557

Updates the "Bid Cycles" menu screen to be the management page for adding/removing the `handshake_allowed_date` for each bid cycle. To test, log in with a user who has superuser access.

### CODE REVIEW

File: src/Components/BidCycleList/BidCycleCard/BidCycleCard.jsx [hunk 1]
```diff
@@ -1,60 +1,136 @@
...
   return item$;
 };
 
-const BidCycleCard = (props) => {
-  const cycle = {
-    name: getCycleInfo(props.cycle, 'name'),
-    start: getCycleInfo(props.cycle, 'cycle_start_date'),
-    end: getCycleInfo(props.cycle, 'cycle_end_date'),
-    deadline: getCycleInfo(props.cycle, 'cycle_deadline_date'),
-  };
```
Comment (elizabeth-jimenez) [inline src/Components/BidCycleList/BidCycleCard/BidCycleCard.jsx:32]: Should we update the default props since these are gone now?

**Comment by elizabeth-jimenez** [comment]
Might be nice to have a spinner while it loads; i thought it was erroring out 😁

**Comment by elizabeth-jimenez** [comment]
image
a little cut-off (1920*n)

File: src/Components/BidCycleList/BidCycleCard/BidCycleCard.jsx [hunk 2]
```diff
@@ -1,60 +1,136 @@
...
+class BidCycleCard extends Component {
+  constructor(props) {
+    super(props);
+    this.state = {
+      editActive: false,
+      isLoading: false,
+      hasErrored: false,
+      data: null,
+      date: props.cycle.handshake_allowed_date || new Date(),
```
Comment (elizabeth-jimenez) [inline src/Components/BidCycleList/BidCycleCard/BidCycleCard.jsx:39]: thought i was seeing double 👭 👬

**Comment by elizabeth-jimenez** [comment]
image
I'm not sure how familiar the users are with the process, but if not very, do you think adding something like this could be helpful?

File: src/actions/bidCycles.js [hunk 3]
```diff
@@ -7,11 +9,29 @@ const receiveBidCycles = data => ({
...
+        results
+          .filter(r => r.state === 'fulfilled' && r.value)
+          .map(m => get(m, 'value.data.results') || m.value.data || []);
+      results$ = results$
+        .map((a, i) =>
+          a
+            .map(m => ({ ...m, _id: i === 1 ? get(m, 'id') : null }))
+            .map((m => ({ ...m, id: m.cycle_id ? `${m.cycle_id}` : `${m.id}` }))),
+        );
+      results$ = values(merge(keyBy(results$[1], 'id'), keyBy(results$[0].map(m => pickBy(m, identity)), 'id')));
```
Comment (elizabeth-jimenez) [inline src/actions/bidCycles.js:32]: Wow, im stealing this 😍

File: src/Components/BidCycleList/BidCycleCard/BidCycleCard.jsx [hunk 4]
```diff
@@ -1,60 +1,136 @@
...
+            hasErrored: false,
+            editActive: false,
+          });
+        })
+        .catch(() => {
+          this.setState({ isLoading: false, hasErrored: true });
+        });
+    };
+
+    const time = new Date(date || new Date());
```
Comment (elizabeth-jimenez) [inline src/Components/BidCycleList/BidCycleCard/BidCycleCard.jsx:94]: Could this just be `new Date(date)`? Or is it catching an edge case outside of line 39?

File: src/Components/BidCycleList/BidCycleCard/BidCycleCard.jsx [hunk 5]
```diff
@@ -1,60 +1,136 @@
...
-    name: NO_BID_CYCLE,
-    cycle_start_date: NO_DATE,
-    cycle_end_date: NO_DATE,
-    cycle_deadline_date: NO_DATE,
+    id: '',
+    name: NO_NAME,
+    handshake_allowed_date: '',
   };
 
   let item$ = get(item, key, defaults[key]);
```
Comment (elizabeth-jimenez) [inline src/Components/BidCycleList/BidCycleCard/BidCycleCard.jsx:22]: I know you didn't write this one, but `let item$ = get(item, key) || defaults[key];`

File: src/Components/BidCycleList/BidCycleCard/BidCycleCard.jsx (hunk 4 above)
Comment (mjoyce91) [reply src/Components/BidCycleList/BidCycleCard/BidCycleCard.jsx]: Good point, safety check already performed

File: src/Components/BidCycleList/BidCycleCard/BidCycleCard.jsx (hunk 5 above)
Comment (mjoyce91) [reply src/Components/BidCycleList/BidCycleCard/BidCycleCard.jsx:22]: There might be a reason for this one

File: src/Components/BidCycleList/BidCycleCard/BidCycleCard.jsx (hunk 4 above)
Comment (elizabeth-jimenez) [reply src/Components/BidCycleList/BidCycleCard/BidCycleCard.jsx]: woot!

File: src/Components/BidCycleList/BidCycleCard/BidCycleCard.jsx (hunk 5 above)
Comment (elizabeth-jimenez) [reply src/Components/BidCycleList/BidCycleCard/BidCycleCard.jsx:22]: ooh

File: src/Components/BidCycleList/BidCycleCard/BidCycleCard.jsx [hunk 6]
```diff
@@ -1,60 +1,136 @@
...
+
+    const cycle$ = {
+      id: getCycleInfo(cycle$$, 'id'),
+      name: getCycleInfo(cycle$$, 'name'),
+      deadline: getCycleInfo(cycle$$, 'handshake_allowed_date'),
+    };
+
+    const patch = (isClear = false) => {
+      this.setState({ isLoading: true, hasErrored: false });
+      const method = cycle$$._id || data?.id ? 'patch' : 'post';
```
[the remark this answers is not in the document]
Comment (mjoyce91) [reply src/Components/BidCycleList/BidCycleCard/BidCycleCard.jsx:74]: Basically the native JS way to do safe nested property checks (i.e. `get`)

**Review by elizabeth-jimenez — APPROVED** [review]
🍰 🧁 👍

File: src/Components/BidCycleList/BidCycleCard/BidCycleCard.jsx (hunk 6 above)
[the remark this answers is not in the document]
Comment (elizabeth-jimenez) [reply src/Components/BidCycleList/BidCycleCard/BidCycleCard.jsx:74]: o didn't know that. Very nice

**Pull request merged**